File: C_Sharp/1001-2000/p1657.cs
namespace p1657 {
    public class Solution {
        public bool CloseStrings(string word1, string word2) {
            if(word1.Length != word2.Length){
                return false;
            }
            int [] w1 = new int [26];
            int [] w2 = new int [26];

            for(int i=0; i < word1.Length; i++){
                w1[word1[i]-'a']++;
                w2[word2[i]-'a']++;
            }

            for(int i = 0; i < 26; i++){
                if(w1[i] > 0 && w2[i] == 0){
                    return false;
                }
                if(w2[i] > 0 && w2[i] == 0){
                    return false;
                }
            }

            Array.Sort(w1);
            Array.Sort(w2);
            return w1.SequenceEqual(w2);
        }
    }    
}

File: C_Plus_Plus/1001-2000/p1657.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    /**
     * @brief For this problem we can freely rearrange the string characters and
     * swap all occurrences of a character with the instances of another character
     * 
     * @param word1 
     * @param word2 
     * @return true 
     * @return false 
     */
    bool closeStrings(string word1, string word2) {
        if(word1.size() != word2.size()){
            return false;
        }
        vector<int> w1(26,0);
        vector<int> w2(26,0);
        //creating dictionaries showing the frequency of each character
        for(int i=0; i < word1.size(); i++){
            w1[word1[i]-'a']++;
            w2[word2[i]-'a']++;
        }
        //checking that both strings use the same sets of characters
        for(int i = 0; i < 26; i++){
            if(w1[i] > 0 && w2[i] == 0){
                return false;
            }
            if(w2[i] > 0 && w2[i] == 0){
                return false;
            }
        }
        //If each string has the same number of occurrences for a different chars
        //we can swap occurrences to make them match
        sort(w1.begin(), w1.end());
        sort(w2.begin(), w2.end());
        for(int i = 0; i < 26; i++){
            if(w1[i] != w2[i]){
                return false;
            }
        }
        return true;
    }
};